identify.1
.ad l
.nh
.TH identify 1 "1 May 1994" "ImageMagick"
.SH NAME
identify - describe the format and characteristics of one or more image
files.
.SH SYNOPSIS
.B "identify"
[ \fIoptions\fP ... ] \fIfile\fP [ \fIfile\fP ... ]
.SH DESCRIPTION
\fBidentify\fP describes the format and characteristics of one or more image
files.  It will also report if an image is incomplete or corrupt.
The information displayed includes the scene number, the file name, the
width and height of the image, whether the image is colormapped or not,
the number of colors in the image, the number of bytes in the image, the
format of the image (JPEG, PNM, etc.), and finally the number of seconds
it took to read and process the image.  An example line output from
\fBidentify\fP follows:

.nf
     images/aquarium.miff 640x480 PseudoClass 256c 308135b MIFF 1s
.fi

If \fB-verbose\fP is set, expect additional output including any image
comment:

.nf
     Image: images/aquarium.miff
       class: PseudoClass
       colors: 256
       signature: eb5dca81dd93ae7e6ffae99a5275a53e
       matte: False
       geometry: 640x480
       depth: 8
       bytes: 308135
       format: MIFF
       comments:

         Imported from MTV raster image:  aquarium.mtv
.fi
.SH OPTIONS
.TP
.B "-cache \fIthreshold\fP"
megabytes of memory available to the pixel cache.

Image pixels are stored in memory until 80 megabytes of memory
have been consumed.  Subsequent pixel operations are cached on disk.
Operations to memory are significantly faster but if your computer does not
have a sufficient amount of free memory you may want to adjust this threshold
value.
.TP
.B "-ping"
efficiently determine image characteristics.

This is a more efficient and less memory intensive way to query if an image
exists and what its size is.  Note, only the size of the first image in
a multi-frame image file is returned.
.TP
.B "-size \fI<width>x<height>+<offset>\fP"
width and height of the image.

Use this option to specify the width and height of raw images whose
dimensions are unknown such as \fBGRAY\fP, \fBRGB\fP, or \fBCMYK\fP.
In addition to width and height, use \fB-size\fP to skip any header
information in the image or tell the number of colors in a \fBMAP\fP
image file, (e.g. -size 640x512+256).

For Photo CD images, choose from these sizes:

.nf
      192x128
      384x256
      768x512
     1536x1024
     3072x2048
.fi

Finally, use this option to choose a particular resolution layer of a JBIG
or JPEG image (e.g. -size 1024x768).
.TP
.B -verbose
print detailed information about the image.

This information is printed:  image name;  the image class
(\fIDirectClass\fP or \fIPseudoClass\fP);  the total number of unique
colors; whether there is a matte associated with the image; the number
of runlength packets; the image size; the depth of the image; the image
format; the image scene; and finally any image comment.  Refer to
\fBmiff(1)\fP for a description of the image class.
.SH SEE ALSO
.B
display(1), animate(1), montage(1), mogrify(1), convert(1), combine(1), xtp(1)
.SH COPYRIGHT
1998 1998 E. I. du Pont de Nemours and Company

Permission is hereby granted, free of charge, to any person obtaining a
copy of this software and associated documentation files ("ImageMagick"),
to deal in ImageMagick without restriction, including without limitation
the rights to use, copy, modify, merge, publish, distribute, sublicense,
and/or sell copies of ImageMagick, and to permit persons to whom the
ImageMagick is furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in
all copies or substantial portions of ImageMagick.

The software is provided "as is", without warranty of any kind, express or
implied, including but not limited to the warranties of merchantability,
fitness for a particular purpose and noninfringement.  In no event shall
E. I. du Pont de Nemours and Company be liable for any claim, damages or
other liability, whether in an action of contract, tort or otherwise,
arising from, out of or in connection with ImageMagick or the use or other
dealings in ImageMagick.

Except as contained in this notice, the name of the E. I. du Pont de
Nemours and Company shall not be used in advertising or otherwise to
promote the sale, use or other dealings in ImageMagick without prior
written authorization from the E. I. du Pont de Nemours and Company.
.SH AUTHORS
John Cristy, E.I. du Pont De Nemours and Company Incorporated


NEWS
This file was last updated on April 1, 2009.

Significant changes associated with GraphicsMagick 1.1.15 (April 1, 2009)

  Bugs Fixed:

    o The MagickLibVersion define overflowed as of 1.1.10 since only
      one digit was reserved for the minor version.  Fixed now.

    o MSL: Fix for handling of geometry x,y values in association with
      gravity.

    o TIFF: Ping mode is now fast like it should have been.

    o Conjure subcommand was returning inverted status to the
      environment.

    o Benchmark subcommand was leaking memory for each iteration.

    o Magick++ Image color quantization error measurement was
      performing a wasteful operation.

    o GraphicsMagick.spec was not installing module .la files by default.

    o Bounding box query was failing rather than returning a bounding box
      matching the entire image if there was no subimage to bound.

    o Memory leak/use fixes.

  Security Fixes:

    o Exclude all virtual delegates and coders for pseudo-formats when
      deciding to use file extension as the format.

    o Fix assertion thrown for BMP and DIB files with negative image
      height values.

Significant changes associated with GraphicsMagick 1.1.14 (released June 5, 2008)

  Bugs Fixed:

    o Noise generation was not working correctly for anything but a Q8 build.

    o Poisson noise generator was extremely slow.  Now it is reasonably fast.

    o PDF reader now computes bounding box for rotated PDFs correctly.

  Security Fixes:

    o Cineon reader: Fixed crash with broken file from Sami Liedes.

    o PICT reader: Fixed crash with broken files from Sami Liedes.

    o XCF reader: Fixed crash with broken files from Sami Liedes.

  Security Issue Not Fixed:

    o JP2/JPC reader: Several JPEG-2000 files from Sami Liedes broken collection
      cause crashes in Jasper.  This is not GraphicsMagick's fault.

Significant changes associated with GraphicsMagick 1.1.13 (released May 28, 2008)

  Bugs Fixed:

    o Documentation is now installed according to the conventions established by
      the configure script.  This results in documentation being installed under
      /usr/local/share/doc/GraphicsMagick by default.

    o HWB colorspace now works correctly.

    o Composition with CopyOpacity now produces an image with
      transparency (as expected).

    o Composition now preserves the canvas colorspace.

    o Composition with a displacement map (-displace) no longer leaks an image.

    o Alpha composition now works as expected when both pixels involved
      include transparency.

    o -gamma multiple channel syntax now works as documented.

    o Now compiles with Visual C++ 2008.

Significant changes associated with GraphicsMagick 1.1.12 (released April 28, 2008)

  Security Fixes:

    o Do not access X11 or invoke convenience or stealth delegate
      programs based on the file extension. In particular, these file
      extensions are rejected for consideration as a format specifier:
      'autotrace', 'browse', 'dcraw', 'edit', 'gs-color',
      'gs-color+alpha', 'gs-gray', 'gs-mono', 'launch', 'mpeg-encode',
      'print', 'scan', 'show', 'win', 'xc', and 'x'.

  Bugs Fixed:

    o magick/effect.c: Should now compile for ARM CPU.

    o TIFF: Don't request Kodak private tags since these cause some
      versions of libtiff to misbehave.

    o When performing string expansion of image attribute identifiers, skip
      those which require access to image pixels if image pixels are not
      present.

    o CropImageToHBITMAP(), ImageToHBITMAP(): Fix leak of bitmap handle.

Significant changes associated with GraphicsMagick 1.1.11 (released February 23, 2008)

  Bugs Fixed:

    o BMP: Support large files.

    o DIB: Support large files.

    o PNG: Fix depth handling with 16-bit PNG files in the Q8 build.

    o SUN: Properly report image depth.

    o TIFF: Endian option (-endian) now controls TIFF byte endian order
      rather than bit fill order.

    o DCM, DIB, XBM, XCF, XWD: Eliminate integer overflow vulnerability
      (IDefense 09.19.07).

    o HSL colorspace transform: Avoid optimization bug noticed on Opteron
      with GCC.

    o HWB colorspace transform: Avoid optimization bug noticed on Opteron
      with GCC.

    o RGBTransformImage()/TransformRGBImage(): Was using HWB colorspace
      when HSL was requested.

    o Successfully reads files with names like 'file[123]'.

    o 'gm display': No longer rely on isatty() to determine if input is
      from a pipe (use 'gm display -' to display an image read from a pipe).

  Feature Improvements:

    o 'identify +ping' forces the pixels to be read (similar to GM 1.2).

    o 'gm -version' now indicates if build supports "Large Memory" (i.e. 64-bit).

    o TIFF: Use '-define tiff:fill-order={msb2lsb|lsb2msb}' to control
      TIFF bit fill order.


  Performance Improvements:

    o No longer bogs down if a directory contains hundreds of thousands of
      files and the filename looks like a wildcard specification.


Significant changes associated with GraphicsMagick 1.1.10 (released September 19, 2007)

  Bugs Fixed:

    o Image rotate by -90/270 degrees was producing wrong output.

Significant changes associated with GraphicsMagick 1.1.9 (released September 16, 2007)

  Bugs Fixed:

    o In mogrify command, don't remove file name based on random junk in
      memory.
    o Fixed memory leak when reading MPC files.
    o Fixed crash when writing MIFF format and depth is not expected 8/16/32/.
    o Find latest Ghostscript under Windows which identifies itself as
      "GPL Ghostscript".
    o In mogrify command, don't leak memory in the case where the image
      file contains multiple frames.
    o Fixed crash in PNG and JPEG coders when the image to be written is
      part of an image list.
    o PNG reader errors are now properly reported to the user.
    o TIFF output can now be written to a pipe or other non-seekable
      destination.

utilities/identify.c
/*
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%                                                                             %
%                                                                             %
%                                                                             %
%           IIIII  DDDD   EEEEE  N   N  TTTTT  IIIII  FFFFF  Y   Y            %
%             I    D   D  E      NN  N    T      I    F       Y Y             %
%             I    D   D  EEE    N N N    T      I    FFF      Y              %
%             I    D   D  E      N  NN    T      I    F        Y              %
%           IIIII  DDDD   EEEEE  N   N    T    IIIII  F        Y              %
%                                                                             %
%                                                                             %
%               Identify an Image Format and Characteristics.                 %
%                                                                             %
%                                                                             %
%                                                                             %
%                           Software Design                                   %
%                             John Cristy                                     %
%                            September 1994                                   %
%                                                                             %
%                                                                             %
%  Copyright (C) 2001 ImageMagick Studio, a non-profit organization dedicated %
%  to making software imaging solutions freely available.                     %
%                                                                             %
%  Permission is hereby granted, free of charge, to any person obtaining a    %
%  copy of this software and associated documentation files ("ImageMagick"),  %
%  to deal in ImageMagick without restriction, including without limitation   %
%  the rights to use, copy, modify, merge, publish, distribute, sublicense,   %
%  and/or sell copies of ImageMagick, and to permit persons to whom the       %
%  ImageMagick is furnished to do so, subject to the following conditions:    %
%                                                                             %
%  The above copyright notice and this permission notice shall be included in %
%  all copies or substantial portions of ImageMagick.                         %
%                                                                             %
%  The software is provided "as is", without warranty of any kind, express or %
%  implied, including but not limited to the warranties of merchantability,   %
%  fitness for a particular purpose and noninfringement.  In no event shall   %
%  ImageMagick Studio be liable for any claim, damages or other liability,    %
%  whether in an action of contract, tort or otherwise, arising from, out of  %
%  or in connection with ImageMagick or the use or other dealings in          %
%  ImageMagick.                                                               %
%                                                                             %
%  Except as contained in this notice, the name of the ImageMagick Studio     %
%  shall not be used in advertising or otherwise to promote the sale, use or  %
%  other dealings in ImageMagick without prior written authorization from the %
%  ImageMagick Studio.                                                        %
%                                                                             %
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%
%  Identify describes the format and characteristics of one or more image
%  files.  It will also report if an image is incomplete or corrupt.
%
%  The identify program syntax is:
%
%  Usage: identify [options ...] file [ files... ]
%
%  Where options include:
%    -cache threshold  megabytes of memory available to the pixel cache
%    -density geometry vertical and horizontal density of the image
%    -format "string"  output formatted image characteristics
%    -ping             efficiently determine image characteristics
%    -size geometry    width and height of image
%    -verbose          print detailed information about the image
%
%
*/

/*
  Include declarations.
*/
#include "magick/magick.h"
#include "magick/defines.h"

/*
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%                                                                             %
%                                                                             %
%                                                                             %
%   U s a g e                                                                 %
%                                                                             %
%                                                                             %
%                                                                             %
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%
%  Procedure Usage displays the program usage;
%
%  The format of the Usage method is:
%
%      void Usage()
%
%
*/
static void Usage()
{
  const char
    **p;

  static const char
    *options[]=
    {
      "-cache threshold  megabytes of memory available to the pixel cache",
      "-density geometry vertical and horizontal density of the image",
      "-format \"string\"  output formatted image characteristics",
      "-ping             efficiently determine image characteristics",
      "-size geometry    width and height of image",
      "-verbose          print detailed information about the image",
      (char *) NULL
    };

  unsigned int
    version;

  (void) printf("Version: %.1024s\n",GetMagickVersion(&version));
  (void) printf("Copyright: %.1024s\n\n",MagickCopyright);
  (void) printf("Usage: %.1024s [ options ... ] file [ file ... ]\n",
    SetClientName((char *) NULL));
  (void) printf("\nWhere options include:\n");
  for (p=options; *p != (char *) NULL; p++)
    (void) printf("  %.1024s\n",*p);
  Exit(0);
}

/*
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%                                                                             %
%                                                                             %
%                                                                             %
%    M a i n                                                                  %
%                                                                             %
%                                                                             %
%                                                                             %
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%
%
*/
int main(int argc,char **argv)
{
  char
    *format,
    *option;

  double
    sans;

  ExceptionInfo
    exception;

  Image
    *image,
    *p;

  ImageInfo
    *image_info;

  int
    number_images;

  register int
    i;

  unsigned int
    count,
    status;

  /*
    Initialize command line arguments.
  */
  ReadCommandlLine(argc,&argv);
  MagickIncarnate(*argv);
  format=(char *) NULL;
  for (i=1; i < argc; i++)
  {
    option=argv[i];
    if (LocaleNCompare("-format",argv[i],2) == 0)
      {
        i++;
        if (i == argc)
          MagickError(OptionError,"Missing format string",option);
        CloneString(&format,argv[i]);
        argv[i]=(char *) "";
        break;
      }
  }
  status=ExpandFilenames(&argc,&argv);
  if (status == False)
    MagickError(ResourceLimitError,"Memory allocation failed",(char *) NULL);
  if (argc < 2)
    Usage();
  /*
    Set defaults.
  */
  count=0;
  GetExceptionInfo(&exception);
  image_info=CloneImageInfo((ImageInfo *) NULL);
  number_images=0;
  /*
    Identify an image.
  */
  for (i=1; i < argc; i++)
  {
    option=argv[i];
    if ((Extent(option) >= 2) && ((*option == '-') || (*option == '+')))
      {
        switch(*(option+1))
        {
          case 'c':
          {
            if (LocaleNCompare("cache",option+1,3) == 0)
              {
                if (*option == '-')
                  {
                    i++;
                    if ((i == argc) || !sscanf(argv[i],"%lf",&sans))
                      MagickError(OptionError,"Missing threshold",option);
                  }
                SetCacheThreshold(atoi(argv[i]));
                break;
              }
            MagickError(OptionError,"Unrecognized option",option);
            break;
          }
          case 'd':
          {
            if (LocaleNCompare("debug",option+1,3) == 0)
              {
                image_info->debug=(*option == '-');
                break;
              }
            if (LocaleNCompare("density",option+1,3) == 0)
              {
                CloneString(&image_info->density,(char *) NULL);
                if (*option == '-')
                  {
                  i++;
                    if ((i == argc) || !IsGeometry(argv[i]))
                      MagickError(OptionError,"Missing geometry",option);
                    (void) CloneString(&image_info->density,argv[i]);
                  }
                break;
              }
            MagickError(OptionError,"Unrecognized option",option);
            break;
          }
          case 'f':
          {
            if (LocaleNCompare("format",option+1,2) == 0)
              {
                if (*option == '-')
                  {
                    i++;
                    if (i == argc)
                      MagickError(OptionError,"Missing format string",option);
                  }
                break;
              }
            MagickError(OptionError,"Unrecognized option",option);
            break;
          }
          case 'h':
          {
            if (LocaleNCompare("help",option+1,2) == 0)
              {
                Usage();
                break;
              }
            MagickError(OptionError,"Unrecognized option",option);
            break;
          }
          case 'p':
          {
            if (LocaleNCompare("ping",option+1,2) == 0)
              {
                image_info->ping=(*option == '-');
                break;
              }
            MagickError(OptionError,"Unrecognized option",option);
            break;
          }
          case 's':
          {
            if (LocaleNCompare("size",option+1,2) == 0)
              {
                CloneString(&image_info->size,(char *) NULL);
                if (*option == '-')
                  {
                    i++;
                    if ((i == argc) || !IsGeometry(argv[i]))
                      MagickError(OptionError,"Missing geometry",option);
                    (void) CloneString(&image_info->size,argv[i]);
                  }
                break;
              }
            MagickError(OptionError,"Unrecognized option",option);
            break;
          }
          case 'v':
          {
            if (LocaleNCompare("verbose",option+1,2) == 0)
              {
                image_info->verbose=(*option == '-');
                break;
              }
            MagickError(OptionError,"Unrecognized option",option);
            break;
          }
          case '?':
          {
            Usage();
            break;
          }
          default:
          {
            MagickError(OptionError,"Unrecognized option",option);
            break;
          }
        }
        continue;
      }
    /*
      Identify image.
    */
    (void) strcpy(image_info->filename,argv[i]);
    if (image_info->ping)
      {
        /*
          Get first frame only.
        */
        image_info->verbose=False;
        image_info->subimage=0;
        image_info->subrange=0;
        image=ReadImage(image_info,&exception);
        if (image == (Image *) NULL)
          {
            MagickWarning(exception.severity,exception.reason,
              exception.description);
            continue;
          }
        if (format == (char *) NULL)
          DescribeImage(image,stdout,False);
        else
          {
            char
              *text;

            text=TranslateText((ImageInfo *) NULL,image,format);
            if (text == (char *) NULL)
              ThrowBinaryException(ResourceLimitWarning,
                "Unable to format image data","Memory allocation failed");
            (void) fputs(text,stdout);
          }
        number_images++;
        continue;
      }
    image=ReadImage(image_info,&exception);
    if (image == (Image *) NULL)
      {
        MagickWarning(exception.severity,exception.reason,exception.description);
        continue;
      }
    for (p=image; p != (Image *) NULL; p=p->next)
    {
      if (p->scene == 0)
        p->scene=count++;
      if (format == (char *) NULL)
        DescribeImage(p,stdout,image_info->verbose);
      else
        {
          char
            *text;

          text=TranslateText((ImageInfo *) NULL,image,format);
          if (text == (char *) NULL)
            ThrowBinaryException(ResourceLimitWarning,
              "Unable to format image data","Memory allocation failed");
          (void) fputs(text,stdout);
          (void) fputc('\n',stdout);
        }
    }
    DestroyImages(image);
    number_images++;
  }
  if ((i != argc) || (number_images == 0))
    MagickError(OptionError,"Missing an image file name",(char *) NULL);
  DestroyImageInfo(image_info);
  LiberateMemory((void **) &argv);
  Exit(0);
  return(False);
}
